added ssh_keys table for id_rsa files.

nerddotcat opened the pull request:
Added a new table for tracking id_rsa file. I'm new to contributing to open source software and I tried to follow the contribution guide as closely as possible. So let me know if there is anything I need to add or change.

facebook-github-bot-4 commented:
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

theopolis commented on osquery/tables/system/ssh_keys.cpp:
```diff
@@ -0,0 +1,59 @@
+namespace osquery {
+namespace tables {
+
+const std::vector<std::string> kSSHUserKeys = {".ssh/id_rsa"};
```
You could also look at *all* files within each user's `~/.ssh` directory.

theopolis commented on osquery/tables/system/ssh_keys.cpp:
```diff
@@ -0,0 +1,59 @@
+#include <string>
```
This file will need the default copyright, check out the other table files.

theopolis commented on osquery/tables/system/ssh_keys.cpp:
```diff
@@ -0,0 +1,59 @@
+#include <osquery/tables.h>
+#include <osquery/filesystem.h>
+
+#include "osquery/core/conversions.h"
```
Are you using any methods from these local includes: `conversions.h` and `system_utils.h`?

theopolis commented on osquery/tables/system/ssh_keys.cpp:
```diff
@@ -0,0 +1,59 @@
+      r["is_encrypted"] = INTEGER(1);
+    }
+    results.push_back(r);
+    
```
nit, can you delete these unneeded spaces?

theopolis commented:
Thanks for the new table @nerddotcat! If you have the `clang-format` application (script rather) you can:

```
$ git add .
$ make format
```

And it will fix some minor style edits too.

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

PickmanSec commented:
Could it also possibly include ssh pub key fingerprints?

facebook-github-bot-2 commented:
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

nerddotcat commented:
Thanks for all the feedback on my code, I'll work to fix all of these issues!

nerddotcat commented:
Made a new commit for this pull request, looks like it gets added automatically?

- Added the default copyright to the ssh_keys.cpp file
- Removed the conversions.h include because it wasn't being used and system_utils.h is being used because of the usersFromContext method
- Used make format to remove unnecessary spaces
- Now checks all files in the ~/.ssh directory and looks for the string "PRIVATE KEY" to check if the file is a private key
- Changed the name of the column "is_encrypted" to "encrypted"

I looked into adding public key info/fingerprints but I'm not sure how I should do that at this time.

theopolis commented:
Right, Github is using your remote branch `nerddotcat:id_rsa-table` for the pull request. If you remove commits, add commits, or force push to overwrite commits `-f` to that branch, the pull request will be notified by Github's postcommit hooks. For the osquery project, if build/test is enabled for your pull request, that will trigger again too.

theopolis commented on osquery/tables/system/ssh_keys.cpp:
```diff
@@ -0,0 +1,77 @@
+void genSSHkeyForHosts(const std::string& uid,
+                        const std::string& directory,
+                        QueryData& results) {
+  //get list of files in directory
```
nitpicking: your comment style is inconsistent. Either capitalize the start of each, or don't, but please don't mix. Also, it's best to begin each with a space after the double-slash: `// `.

theopolis commented:
Looking good, see my inline comment about comments.

And do you have `clang-format` installed? Would you mind:
```bash
clang-format -i ./osquery/tables/system/ssh_keys.cpp
```

theopolis commented:
ok to test

osqueryer commented:
Build origin/pr/2245/merge (Job: 2887) is good! :+1:

nerddotcat commented:
I might have forgot to commit my change when I first used make format, but I think everything is good now. Also updated the comments to use the correct format.

osqueryer commented:
Build origin/pr/2245/merge (Job: 2889) is good! :+1:

PoppySeedPlehzr commented on specs/posix/ssh_keys.table:
```diff
@@ -0,0 +1,15 @@
+schema([
+    Column("uid", BIGINT, "The local user that owns the ssh_keys file",
+      additional=True),
+    Column("key_file", TEXT, "Path to key file"),
```
Would you mind making `path` instead of `key_file`? With `path` we could join against other tables in osquery.

PoppySeedPlehzr commented on osquery/tables/system/ssh_keys.cpp:
```diff
@@ -0,0 +1,75 @@
+      // File is private key, create record for it
+      Row r;
+      r["uid"] = uid;
+      r["key_file"] = kfile;
```
Same as below.

PoppySeedPlehzr commented:
Would it be possible to also grab public keys? The table name `ssh_keys` gives me the impression it'll be all keys stored under the `.ssh` directory.

apfelwerk commented:
+1

> On 18 Jul 2016, at 16:30, Nick Anderson <notifications@github.com> wrote:
> 
> Would it be possible to also grab public keys? The table name ssh_keys gives me the impression it'll be all keys stored under the .ssh directory.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

osqueryer commented:
Build origin/pr/2245/merge (Job: 2890) is good! :+1:

nerddotcat commented:
Renamed table to user_ssh_keys.
Renamed key_file to path.

I can add public key information but I have a question. Is it fine to assume that all public key files have the same name of the private key, but with a .pub suffix? Or is there a more proper way of finding the public key file?

PoppySeedPlehzr commented:
@nerddotcat it'd be best not to make any assumptions at the file name, however that brings up issue with other files residing in the `.ssh` directory. Mayhaps we need to think on that issue and proceed with just getting your priv keys table rolled in :)
